### ISSUE TITLE
[FR] Custom path for artifacts in backend store on the run level

### ISSUE THREAD

**Issue by lvijnck** [mlflow/mlflow#13016]
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

MLFlow uses a path akin to `<artifact_root>/<num>/<run_id>/artifacts/<artifact_artifact_path>`. It would be nice to supply a custom root for a run when using MLFlow.``

### Motivation

> #### What is the use case for this feature?

More control of where files are materialized

> #### Why is this use case valuable to support for MLflow users in general?

More control of where files are materialized

> #### Why is this use case valuable to support for your project(s) or organization?

MLFlow is only managing a subset of our files, we would like more control over the file locations in the backend.

> #### Why is it currently difficult to achieve this use case?

We can use the `artifact_path` on the experiment level, but not on the run level.

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by harupy** [comment]
@lvijnck Thanks for the FR.

> we would like more control over the file locations in the backend.

Can you elaborate on this?

**Comment by lvijnck** [comment]
Hi @harupy,

We're using the `--no-serve-artifacts` option for the tracking server, and use GCS as a backend. We're using MLFlow in Combination with Kedro and have a well defined directory structure for our pipeline outputs (combination of Spark, Pandas and MathPlotLib datasets). 

The root directory for a run is a as follows:

```
gs://<bucket_name>/data/run/<kedro-run-id>
```

(not the kedro-run-id is _not_ the MLFlow run id)

We would hence like to store the files from MLFlow under the same directory, unfortunately this is not possible due to MLFlow not allowing to specify an artifact root path on the run level (it's available on the experiment level).

In summary, I don't think MLFlow should manage the layout of files on object storage. The responsibility of MLFlow should be to index the files and make them comparable.
